diagnostic_aggregator/src/aggregator.cpp

```diff
@@ -107,13 +106,77 @@ void Aggregator::diagCallback(const diagnostic_msgs::DiagnosticArray::ConstPtr&
   }
 }
 
-Aggregator::~Aggregator() 
+Aggregator::~Aggregator()
 {
   if (analyzer_group_) delete analyzer_group_;
 
   if (other_analyzer_) delete other_analyzer_;
 }
 
+
+void Aggregator::bondBroken()
+{
+  boost::mutex::scoped_lock lock(mutex_); // Possibility of multiple bonds breaking at once
+
+  vector<boost::shared_ptr<bond::Bond> >::iterator it;
+  for (it = bonds_.begin(); it != bonds_.end();)
+  {
+    if ((*it)->isBroken())
+    {
+      string id = (*it)->getId();
+      ROS_DEBUG("bond for namespace %s was broken", id.c_str());
+      analyzer_group_->removeAnalyzer(added_analyzers_.find(id)->second);
+      added_analyzers_.erase(id);
+      bonds_.erase(it); // it points to next in list (or end())
+    } else {
+      // increment in afterthought will go past bonds_.end() if erase()ing last
+      // element, so do it here instead
+      ++it;
+    }
+  }
+}
+
+void Aggregator::bondFormed(){}
+
+bool Aggregator::addDiagnostics(diagnostic_msgs::AddDiagnostics::Request &req,
+				diagnostic_msgs::AddDiagnostics::Response &res)
+{
+  // Wait for connections on this bond. Wait until it's formed to put it in the
+  // bond vector
+  ROS_DEBUG("got load request for namespace %s", req.load_namespace.c_str());
+  // rebuff attempts to add things from the same namespace twice
+  if (added_analyzers_.find(req.load_namespace) != added_analyzers_.end()) {
+    res.message = "requested load from namespace " + req.load_namespace + " which is already in use";
+    res.success = false;
+    return true;
+  }
+
+  boost::shared_ptr<bond::Bond> req_bond = boost::make_shared<bond::Bond>(
+    "/diagnostics_agg_bond", req.load_namespace,
+    boost::function<void(void)>(boost::bind(&Aggregator::bondBroken, this)),
+    boost::function<void(void)>(boost::bind(&Aggregator::bondFormed, this))
+    );
+  req_bond->start();
+
+  boost::shared_ptr<Analyzer> group(new AnalyzerGroup());
+
+  bonds_.push_back(req_bond); // bond formed, keep track of it
+
+  if (group->init(base_path_, req.load_namespace))
```

Comment:
I don't think we need a new version of `Analyzer::init()` for this; it's just as easy to create a temporary NodeHandle to specify the namespace here:

```
if (group->init(base_path_, ros::NodeHandle(req.load_namespace)))
```